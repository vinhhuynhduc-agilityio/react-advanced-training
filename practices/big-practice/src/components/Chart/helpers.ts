import {
  FormattedDataTotalTasksByProjects,
  FormattedDataTotalTasksCompleted,
  ProjectsData,
  TaskData
} from '@/types';

// ag-grid
import {
  AgAreaSeriesOptions,
  AgBarSeriesTooltipRendererParams,
  AgChartOptions,
  AgTooltipRendererResult,
} from 'ag-charts-community';

/**
 * Formats the task data for the chart displaying total tasks completed.
 *
 * @param tasks - Array of TaskData containing information about completed tasks.
 * @returns An object containing:
 *   - formattedData: Array of objects representing months with task counts for each year.
 *   - years: Array of unique years extracted from the task data, sorted in ascending order.
 */
const formatDataForChartTotalTasks = (
  tasks: TaskData[]
): { formattedData: FormattedDataTotalTasksCompleted[]; years: number[] } => {
  // Set to store unique years present in the task data
  const yearsSet = new Set<number>();

  // Object to store task counts for each month
  const months: { [key: string]: FormattedDataTotalTasksCompleted } = {
    Jan: { month: 'Jan' },
    Feb: { month: 'Feb' },
    Mar: { month: 'Mar' },
    Apr: { month: 'Apr' },
    May: { month: 'May' },
    Jun: { month: 'Jun' },
    Jul: { month: 'Jul' },
    Aug: { month: 'Aug' },
    Sep: { month: 'Sep' },
    Oct: { month: 'Oct' },
    Nov: { month: 'Nov' },
    Dec: { month: 'Dec' },
  };

  // Process each task to extract the month and year, and update task counts
  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (completedDate !== 'incomplete') {

      // Extract month and year from the completedDate
      const [, monthStr, yearStr] = completedDate.split(' ');
      const currentYear = new Date().getFullYear();
      const currentCentury = Math.floor(currentYear / 100);
      const year = parseInt(`${currentCentury}${yearStr}`, 10); // Convert yearStr to full year (e.g., "23" -> 2023)
      yearsSet.add(year);

      // Initialize the task count for the month and year if not already present
      if (!months[monthStr][year]) {
        months[monthStr][year] = 0;
      }

      // Increment the task count for the respective month and year
      months[monthStr][year]++;
    }
  });

  // Convert the set of years into a sorted array
  const years = Array.from(yearsSet).sort();

  const formattedData = Object.values(months).map((month) => {
    years.forEach((year) => {
      if (month[year] === undefined) {
        month[year] = 0;
      }
    });
    return month;
  });

  // Return the formatted data and list of years
  return { formattedData, years };
};

const formatDataForChartTotalTasksByProjects = (
  tasks: TaskData[],
  projects: ProjectsData[]
): { formattedData: FormattedDataTotalTasksByProjects[]; years: number[] } => {
  const yearsSet = new Set<number>();

  // Initialize the result with all projects, defaulting years to 0
  const result = projects.reduce((acc, project) => {
    acc[project.projectName] = {
      projectName: project.projectName,
    };
    return acc;
  }, {} as { [key: string]: FormattedDataTotalTasksByProjects });

  // Populate task data
  tasks.forEach((task) => {
    const { projectName, startDate } = task;
    const [, , yearStr] = startDate.split(' ');
    const currentYear = new Date().getFullYear();
    const currentCentury = Math.floor(currentYear / 100);
    const year = parseInt(`${currentCentury}${yearStr}`, 10);

    yearsSet.add(year);

    if (!result[projectName][year]) {
      result[projectName][year] = 0; // Initialize year if not already present
    }

    result[projectName][year]++; // Increment the task count
  });

  // Ensure all projects have values for all years
  const years = Array.from(yearsSet).sort();
  const formattedData = Object.values(result).map((project) => {
    years.forEach((year) => {
      if (!project[year]) {
        project[year] = 0; // Default missing years to 0
      }
    });
    return project;
  });

  return { formattedData, years };
};

const formatDataForChartIndividualEmployee = (
  tasks: TaskData[],
  selectedUserId: string
) => {
  const months: Record<string, { month: string; totalTasksCompleted: number }> = {
    Jan: { month: 'Jan', totalTasksCompleted: 0 },
    Feb: { month: 'Feb', totalTasksCompleted: 0 },
    Mar: { month: 'Mar', totalTasksCompleted: 0 },
    Apr: { month: 'Apr', totalTasksCompleted: 0 },
    May: { month: 'May', totalTasksCompleted: 0 },
    Jun: { month: 'Jun', totalTasksCompleted: 0 },
    Jul: { month: 'Jul', totalTasksCompleted: 0 },
    Aug: { month: 'Aug', totalTasksCompleted: 0 },
    Sep: { month: 'Sep', totalTasksCompleted: 0 },
    Oct: { month: 'Oct', totalTasksCompleted: 0 },
    Nov: { month: 'Nov', totalTasksCompleted: 0 },
    Dec: { month: 'Dec', totalTasksCompleted: 0 },
  };

  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (task.userId === selectedUserId && completedDate !== 'incomplete') {
      const [, monthStr] = completedDate.split(' ');

      months[monthStr as keyof typeof months].totalTasksCompleted++;
    }
  });

  return Object.values(months);
};

const renderTooltipChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const month = params.datum[params.xKey];
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: `<div style='text-align: center; line-height: 1.5'>
              <div>${month}</div>
              <div>${tasksCompleted} tasks completed</div>
            </div>`,
    content: '',
    backgroundColor: '#181d1f',
  };
};

// Default chart configuration
const individualEmployeeProgressOptions: AgChartOptions = {
  title: {
    text: `Individual employee's progress`,
  },
  data: [],
  series: [
    {
      type: 'area',
      xKey: 'month',
      yKey: 'totalTasksCompleted',
      yName: 'Loading...',
      interpolation: { type: 'smooth' },
      tooltip: {
        renderer: renderTooltipChart,
      },
    } as AgAreaSeriesOptions,
  ],
  legend: {
    enabled: true,
    listeners: {
      legendItemClick: (event) => {
        event.preventDefault();
      },
    },
  },
};

const renderTooltipProjectChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: tasksCompleted,
    content: '',
    backgroundColor: '#181d1f',
  };
};

export {
  formatDataForChartTotalTasks,
  formatDataForChartTotalTasksByProjects,
  formatDataForChartIndividualEmployee,
  renderTooltipProjectChart,
  renderTooltipChart,
  individualEmployeeProgressOptions,
};
