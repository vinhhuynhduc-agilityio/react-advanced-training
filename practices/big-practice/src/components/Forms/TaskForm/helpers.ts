// types
import {
  TaskData,
} from '@/types';

const formatDropdownOptions = <T extends { id: string }>(
  data: T[],
  valueKey: keyof T
) =>
  data.map((item) => ({
    id: item.id,
    value: item[valueKey],
  }));

/**
 * Checks if a task name already exists in the task list.
 *
 * @param tasks - Array of existing tasks
 * @param value - New task name to check
 * @returns True if the task name already exists, otherwise false
 */
const isTaskDuplicate = (
  tasks: TaskData[],
  value: string
): boolean => {
  return tasks.some(
    task => task.taskName.toUpperCase() === value.toUpperCase()
  );
};

export {
  formatDropdownOptions,
  isTaskDuplicate
};
