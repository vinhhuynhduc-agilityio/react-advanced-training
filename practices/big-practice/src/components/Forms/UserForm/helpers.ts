import { UserData } from "@/types";

const emailValidation = (defaultEmail: string) => {

  const isEmailDuplicate = (email: string, users: UserData[]) => {
    return users.some(
      (user) => user.email === email && user.email !== defaultEmail
    );
  };

  return { isEmailDuplicate };
};

const readFileAsBase64 = (
  file: File,
  onSuccess: (base64: string) => void
) => {
  const reader = new FileReader();

  reader.readAsDataURL(file);

  reader.onloadend = () => {
    onSuccess(reader.result as string);
  };
};

export {
  emailValidation,
  readFileAsBase64
};
