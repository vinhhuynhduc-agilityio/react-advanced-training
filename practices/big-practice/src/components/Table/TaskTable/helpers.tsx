import { FIELD_TYPE, FieldType } from "@/constant";
import { FieldValue, ProjectsData, TaskData, UserData } from "@/types";

const handleTaskNameChange = (
  value: FieldValue,
  row: TaskData,
  originalTaskNameRef: React.MutableRefObject<string>,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void,
) => {
  const currentValue = originalTaskNameRef.current;
  const newValue = value;

  if (newValue === currentValue) return;

  handleSaveSelect(FIELD_TYPE.TASK_NAME as FieldType, newValue, row);
};

const handleProjectChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const currentValue = row.projectId;
  const newValue = (value as ProjectsData).id;

  if (newValue === currentValue) return;

  handleSaveSelect(FIELD_TYPE.PROJECT as FieldType, value, row);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          projectId: newValue,
          projectName: (value as ProjectsData).projectName,
        }
        : task
    )
  );
};

const handleUserChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  updateEarningsForUsers: (oldUserId: string, newUserId: string, currency: number, status: boolean) => void,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const { userId, currency, status } = row;
  const oldUserId = userId;
  const newUserId = (value as UserData).id;

  if (newUserId === oldUserId) return;

  handleSaveSelect(FIELD_TYPE.USER as FieldType, value, row);
  updateEarningsForUsers(oldUserId, newUserId, currency, status);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          userId: newUserId,
          fullName: (value as UserData).fullName,
        }
        : task
    )
  );
};

const handleStatusChange = (
  value: FieldValue,
  row: TaskData,
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>,
  updateEarningsOnStatusChange: (userId: string, currency: number, status: boolean) => void,
  handleSaveSelect: (type: FieldType, value: FieldValue, row: TaskData) => void
) => {
  const status = value as boolean;
  const { currency, userId } = row;
  const completedDate = status
    ? new Date().toLocaleDateString('en-GB', {
      day: '2-digit',
      month: 'short',
      year: '2-digit',
    } as Intl.DateTimeFormatOptions)
    : 'incomplete';

  handleSaveSelect(FIELD_TYPE.STATUS as FieldType, { status, completedDate }, row);
  updateEarningsOnStatusChange(userId, currency, status);

  setTasks((prevTasks) =>
    prevTasks.map((task) =>
      task.id === row.id
        ? {
          ...task,
          status,
          completedDate,
        }
        : task
    )
  );
};

export {
  handleTaskNameChange,
  handleProjectChange,
  handleUserChange,
  handleStatusChange
};
