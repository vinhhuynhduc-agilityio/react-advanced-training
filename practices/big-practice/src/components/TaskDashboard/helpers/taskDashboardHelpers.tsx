import { FieldType } from "@/types/fieldEnums";
import {
  FieldValue,
  ProjectsData,
  StatusWithDate,
  TaskData,
  UserData
} from "@/types/table";

export const getUpdatedRow = (
  type: FieldType,
  value: FieldValue,
  row: TaskData
): TaskData => {
  switch (type) {
    case FieldType.PROJECT: {
      const { projectName, id } = value as ProjectsData;

      return {
        ...row,
        projectName,
        projectId: id,
      };
    }

    case FieldType.USER: {
      const { fullName, id } = value as UserData;

      return {
        ...row,
        fullName,
        userId: id,
      };
    }

    case FieldType.TASK_NAME:

      return {
        ...row,
        taskName: value as string,
      };

    case FieldType.STATUS: {
      const { status, completedDate } = value as StatusWithDate;

      return {
        ...row,
        status,
        completedDate,
      };
    }

    default:
      return row;
  }
};
