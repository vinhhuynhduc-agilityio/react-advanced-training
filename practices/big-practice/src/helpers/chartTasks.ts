import {
  FormattedMonthData,
  FormattedProjectData
} from "@/types/chartTypes";
import {
  ProjectsData,
  TaskData
} from "@/types/table";

// ag-grid
import {
  AgAreaSeriesOptions,
  AgBarSeriesTooltipRendererParams,
  AgChartOptions,
  AgTooltipRendererResult,
} from "ag-charts-community";

/**
 * @returns {FormattedMonthData[]} Array containing task completion data by month,
 * each object contains: `month` (month name), `2023` and `2024` (number of completed tasks)
 */
const formatDataForChartTotalTasks = (
  tasks: TaskData[]
): FormattedMonthData[] => {
  const months = {
    Jan: { month: "Jan", 2023: 0, 2024: 0 },
    Feb: { month: "Feb", 2023: 0, 2024: 0 },
    Mar: { month: "Mar", 2023: 0, 2024: 0 },
    Apr: { month: "Apr", 2023: 0, 2024: 0 },
    May: { month: "May", 2023: 0, 2024: 0 },
    Jun: { month: "Jun", 2023: 0, 2024: 0 },
    Jul: { month: "Jul", 2023: 0, 2024: 0 },
    Aug: { month: "Aug", 2023: 0, 2024: 0 },
    Sep: { month: "Sep", 2023: 0, 2024: 0 },
    Oct: { month: "Oct", 2023: 0, 2024: 0 },
    Nov: { month: "Nov", 2023: 0, 2024: 0 },
    Dec: { month: "Dec", 2023: 0, 2024: 0 },
  };

  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (completedDate !== "incomplete") {
      const [, monthStr, yearStr] = completedDate.split(" ");
      const year = parseInt("20" + yearStr);

      if (year === 2023 || year === 2024) {
        months[monthStr as keyof typeof months][year]++;
      }
    }
  });

  return Object.values(months);
};

const formatDataForChartTotalTasksByProjects = (
  tasks: TaskData[],
  projects: ProjectsData[]
): FormattedProjectData[] => {
  // Initialize the result with all projects from the `projects` parameter, and set 0 for both 2023 and 2024
  const result = projects.reduce((acc, project) => {
    acc[project.projectName] = {
      projectName: project.projectName,
      2023: 0,
      2024: 0,
    };

    return acc;
  }, {} as { [key: string]: FormattedProjectData });

  // Update the number of tasks per year for projects
  tasks.forEach((task) => {
    const {
      projectName,
      startDate
    } = task;
    const year = parseInt(`20${startDate.split(" ")[2]}`, 10);

    if (
      result[projectName] &&
      (year === 2023 || year === 2024)
    ) {
      result[projectName][year]++;
    }
  });

  return Object.values(result);
};

const formatDataForChartIndividualEmployee = (
  tasks: TaskData[],
  selectedUserId: string
) => {
  const months: Record<string, { month: string; totalTasksCompleted: number }> = {
    Jan: { month: "Jan", totalTasksCompleted: 0 },
    Feb: { month: "Feb", totalTasksCompleted: 0 },
    Mar: { month: "Mar", totalTasksCompleted: 0 },
    Apr: { month: "Apr", totalTasksCompleted: 0 },
    May: { month: "May", totalTasksCompleted: 0 },
    Jun: { month: "Jun", totalTasksCompleted: 0 },
    Jul: { month: "Jul", totalTasksCompleted: 0 },
    Aug: { month: "Aug", totalTasksCompleted: 0 },
    Sep: { month: "Sep", totalTasksCompleted: 0 },
    Oct: { month: "Oct", totalTasksCompleted: 0 },
    Nov: { month: "Nov", totalTasksCompleted: 0 },
    Dec: { month: "Dec", totalTasksCompleted: 0 },
  };

  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (task.userId === selectedUserId && completedDate !== "incomplete") {
      const [, monthStr] = completedDate.split(" ");

      months[monthStr as keyof typeof months].totalTasksCompleted++;
    }
  });

  return Object.values(months);
};

const renderTooltipChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const month = params.datum[params.xKey];
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: `<div style="text-align: center; line-height: 1.5">
              <div>${month}</div>
              <div>${tasksCompleted} tasks completed</div>
            </div>`,
    content: "",
    backgroundColor: "#181d1f",
  };
};

// Default chart configuration
const initOptions: AgChartOptions = {
  title: {
    text: `Individual employee's progress`,
  },
  data: [],
  series: [
    {
      type: "area",
      xKey: "month",
      yKey: "totalTasksCompleted",
      yName: "Loading...",
      interpolation: { type: "smooth" },
      tooltip: {
        renderer: renderTooltipChart,
      },
    } as AgAreaSeriesOptions,
  ],
  legend: {
    enabled: true,
    listeners: {
      legendItemClick: (event) => {
        event.preventDefault();
      },
    },
  },
};

const renderTooltipProjectChart = (
  params: AgBarSeriesTooltipRendererParams
): AgTooltipRendererResult => {
  const tasksCompleted = params.datum[params.yKey];

  return {
    title: tasksCompleted,
    content: "",
    backgroundColor: "#181d1f",
  };
};

export {
  formatDataForChartTotalTasks,
  formatDataForChartTotalTasksByProjects,
  formatDataForChartIndividualEmployee,
  initOptions,
  renderTooltipChart,
  renderTooltipProjectChart,
};
