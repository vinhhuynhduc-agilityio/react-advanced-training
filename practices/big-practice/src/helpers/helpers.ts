const formatDropdownOptions = <T extends { id: string }>(
  data: T[],
  valueKey: keyof T
) =>
  data.map((item) => ({
    id: item.id,
    value: item[valueKey],
  }));

export {
  formatDropdownOptions
};
