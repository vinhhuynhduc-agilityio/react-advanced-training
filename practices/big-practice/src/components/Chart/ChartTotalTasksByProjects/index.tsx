import {
  useMemo,
  memo
} from 'react';
import { AgCharts } from 'ag-charts-react';
import { AgChartOptions } from 'ag-charts-community';

// helpers
import { formatDataForChartTotalTasksByProjects, renderTooltipProjectChart } from '@/components/Chart/helpers';

// component
import { Spinner } from '@/components/common';

// types
import { ProjectsData, TaskData } from '@/types';

interface ChartTotalTasksByProjectsProps {
  isLoading: boolean;
  tasks: TaskData[];
  projects: ProjectsData[];
}

export const ChartTotalTasksByProjects: React.FC<ChartTotalTasksByProjectsProps> = memo(
  ({ isLoading, tasks, projects }) => {
    const { formattedData, years } = useMemo(
      () => formatDataForChartTotalTasksByProjects(tasks, projects),
      [tasks, projects]
    );

    const memoizedOptions = useMemo(() => {
      const dynamicSeries = years.map((year) => ({
        type: 'bar',
        xKey: 'projectName',
        yKey: year.toString(),
        yName: year.toString(),
        direction: 'horizontal',
        tooltip: {
          renderer: renderTooltipProjectChart,
        },
      }));

      return {
        title: {
          text: 'Total tasks by projects',
        },
        data: formattedData,
        series: dynamicSeries,
        legend: {
          enabled: true,
          position: 'bottom',
          item: {
            marker: {
              size: 10,
            },
            label: {
              fontWeight: 'bold',
            },
          },
        },
      } as AgChartOptions;
    }, [formattedData, years]);

    if (isLoading) {
      return (
        <div
          className="flex-1 bg-white border border-customBorder"
          role="figure"
          aria-label="Loading chart: Total tasks by projects"
        >
          <Spinner />
        </div>
      );
    }

    return (
      <div
        id="chart-tasks-by-projects"
        role="figure"
        aria-label="Chart showing total tasks by projects"
        className="flex-1 bg-white border border-customBorder"
      >
        <AgCharts options={memoizedOptions} />
      </div>
    );
  }
);
