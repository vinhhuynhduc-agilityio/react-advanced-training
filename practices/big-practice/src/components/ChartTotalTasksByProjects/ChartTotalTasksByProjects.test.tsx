import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

// types
import { TaskData, ProjectsData } from "@/types/table";

// helpers
import { formatDataForChartTotalTasksByProjects } from "@/helpers/chartTasks";

// component
import ChartTotalTasksByProjects from "./ChartTotalTasksByProjects";

// Mock AgCharts component to simulate its rendering
jest.mock("ag-charts-react", () => ({
  AgCharts: () => <div data-testid="total-tasks-chart"></div>,
}));

// Mock helper functions
jest.mock("@/helpers/ChartTasks", () => ({
  ...jest.requireActual("@/helpers/ChartTasks"),
  formatDataForChartTotalTasksByProjects: jest.fn(),
}));

describe("ChartTotalTasksByProjects Component", () => {
  const mockTasks: TaskData[] = [
    {
      "id": "71e564f4-7c18-47f7-89f2-abe4b7ec2854",
      "userId": "d290f1ee-6c54-4b01-90e6-d701748f0851",
      "projectId": "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
      "taskName": "Build test",
      "startDate": "10 Aug 24",
      "completedDate": "15 Nov 24",
      "currency": 2000,
      "status": true,
      "projectName": "Support",
      "fullName": "Joe Bloggs"
    },
    {
      "id": "0f9fd8b0-8f5c-4a25-b301-eebc6d5700d5",
      "userId": "123e4567-e89b-12d3-a456-426614174001",
      "projectId": "e38e56a2-4d3c-469d-8345-45d6b5fae9f9",
      "taskName": "Develop Backend",
      "startDate": "01 Sep 24",
      "completedDate": "15 Nov 24",
      "currency": 5000,
      "status": true,
      "projectName": "Quality Management",
      "fullName": "Bob White"
    },
  ];

  const mockProjects: ProjectsData[] = [
    {
      id: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
      projectName: "Support",
    },
    {
      id: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
      projectName: "Failure Testing",
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the chart component", () => {
    const { container } = render(
      <ChartTotalTasksByProjects
        tasks={mockTasks}
        projects={mockProjects}
        isLoading={false}
      />
    );
    expect(screen.getByTestId("total-tasks-chart")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("calls formatDataForChartTotalTasksByProjects with correct data", () => {
    render(
      <ChartTotalTasksByProjects
        tasks={mockTasks}
        projects={mockProjects}
        isLoading={false}
      />
    );
    expect(formatDataForChartTotalTasksByProjects).toHaveBeenCalledWith(
      mockTasks,
      mockProjects
    );
  });

  it("does not crash if there are no tasks or projects", () => {
    const { container } = render(
      <ChartTotalTasksByProjects
        tasks={[]}
        projects={[]}
        isLoading={false}
      />
    );

    expect(screen.getByTestId("total-tasks-chart")).toBeInTheDocument();
    expect(formatDataForChartTotalTasksByProjects).toHaveBeenCalledWith([], []);
    expect(container).toMatchSnapshot();
  });

  it("renders a spinner when isLoading is true", () => {
    const { container } = render(
      <ChartTotalTasksByProjects
        tasks={[]}
        projects={[]}
        isLoading={true}
      />
    );

    const spinner = container.querySelector(
      ".w-12.h-12.border-4.border-gray-300.border-t-blue-500.rounded-full.animate-spin"
    );
    expect(spinner).toBeInTheDocument();

    const mockedChart = screen.queryByTestId("mocked-chart");
    expect(mockedChart).not.toBeInTheDocument();
  });
});
