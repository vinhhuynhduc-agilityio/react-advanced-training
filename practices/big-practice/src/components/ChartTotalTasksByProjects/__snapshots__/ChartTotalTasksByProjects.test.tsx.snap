// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTotalTasksByProjects Component does not crash if there are no tasks or projects 1`] = `
<div>
  <div
    class="flex-1 bg-white border border-customBorder"
  >
    <div
      data-testid="total-tasks-chart"
    />
  </div>
</div>
`;

exports[`ChartTotalTasksByProjects Component renders the chart component 1`] = `
<div>
  <div
    class="flex-1 bg-white border border-customBorder"
  >
    <div
      data-testid="total-tasks-chart"
    />
  </div>
</div>
`;
